### PR TITLE
Implement Api Key authorisation on requests

### PR DESCRIPTION
**Pull request by codebrain**
It is possible to set both the `Basic` auth and `ApiKey` auth together.

Clearly this doesn't make sense to do, and to me this feels like an exception.

Should we throw an exception in the client? (which would only surface at runtime), or do we apply some kind of precedence in which auth to apply if both are set?

I also left in the `string` override as I think this might be a little easier for someone to use, if they are unable to find our `.CreateSecureString()` extension method.

### CODE REVIEW

File: src/Elasticsearch.Net/Configuration/Security/ApiKeyAuthenticationCredentials.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	/// <summary>
+	/// Credentials for Api Key Authentication
+	/// </summary>
+	public class ApiKeyAuthenticationCredentials : IDisposable
+	{
+		public ApiKeyAuthenticationCredentials()
+		{
+		}
+
+		public ApiKeyAuthenticationCredentials(string id, SecureString apiKey)
```
Comment (Mpdreamz) [inline src/Elasticsearch.Net/Configuration/Security/ApiKeyAuthenticationCredentials.cs]: Create an overload that takes the full base64 encoded apiKey as well I reckon. This is what folks'll typically copy paste.

File: src/Elasticsearch.Net/Connection/HttpWebRequestConnection.cs [hunk 2]
```diff
@@ -258,6 +260,19 @@ protected virtual void SetBasicAuthenticationIfNeeded(HttpWebRequest request, Re
...
+		{
+			// ApiKey auth credentials take the following precedence (highest -> lowest):
+			// 1 - Specified on the request (highest precedence)
+			// 2 - Specified at the global IConnectionSettings level
+
+			string apiKey = null;
+			if (requestData.ApiKeyAuthenticationCredentials != null)
+				apiKey = $"{requestData.ApiKeyAuthenticationCredentials.Id}:{requestData.ApiKeyAuthenticationCredentials.ApiKey.CreateString()}";
+
+			if (!string.IsNullOrWhiteSpace(apiKey))
```
Comment (Mpdreamz) [inline src/Elasticsearch.Net/Connection/HttpWebRequestConnection.cs]: Move base64'ing into `ApiKeyAuthenticationCredentials`

File: src/Elasticsearch.Net/Configuration/Security/ApiKeyAuthenticationCredentials.cs (hunk 1 above)
Comment (codebrain) [reply src/Elasticsearch.Net/Configuration/Security/ApiKeyAuthenticationCredentials.cs]: Will set that as a `SecureString` type

File: src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs [hunk 3]
```diff
@@ -434,6 +436,18 @@ protected ConnectionConfiguration(IConnectionPool connectionPool, IConnection co
...
+		/// <summary>
+		/// Api Key to send with all requests to Elasticsearch
+		/// </summary>
+		public T ApiKeyAuthentication(string id, SecureString apiKey) =>
+			Assign(new ApiKeyAuthenticationCredentials(id, apiKey), (a, v) => a._apiKeyAuthentication = v);
+
+		/// <summary>
+		/// Api Key to send with all requests to Elasticsearch
+		/// </summary>
+		public T ApiKeyAuthentication(string id, string apiKey) =>
```
Comment (russcam) [inline src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs]: Kept string password for BasicAuthentication in 7.x to ease the move from 6.x to 7.x, but wonder if we should we only support `SecureString` for apikey going forward?

File: src/Elasticsearch.Net/Connection/HttpWebRequestConnection.cs [hunk 4]
```diff
@@ -239,7 +240,16 @@ protected virtual void SetProxyIfNeeded(HttpWebRequest request, RequestData requ
...
+		private void SetAuthenticationIfNeeded(RequestData requestData, HttpWebRequest request)
+		{
+			// Api Key authentication takes precedence
+			var apiKeySet = SetApiKeyAuthenticationIfNeeded(request, requestData);
+
+			if (!apiKeySet)
+				SetBasicAuthenticationIfNeeded(request, requestData);
+		}
+
+		protected virtual bool SetBasicAuthenticationIfNeeded(HttpWebRequest request, RequestData requestData)
```
Comment (russcam) [inline src/Elasticsearch.Net/Connection/HttpWebRequestConnection.cs]: public API breaking change going from `void` to `bool`

File: src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs (hunk 3 above)
Comment (codebrain) [reply src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs]: Would prefer to keep supporting `string` for ease of use.

**Pull request merged**